### PR TITLE
[FIX] stock_account: validate picking with branch company

### PR DESCRIPTION
**Pull request by guva-odoo**
We get an error when validating a stock picking with
a branch company

Steps:

- Create a company C with a branch B
- Set env.companies to both C and B, and env.company to either C or B
- Create product P with company set as B, a standard_price > 0 and a
  product category set to 'Automated' and valuation != 'standard'
- Make a stock picking from/to B for product P
- Try to validate picking
-> UserError: 'Incompatible companies on records'

This is because the account move is created with main company instead
of branch company.

opw-4105551

### CODE REVIEW

File: addons/stock_account/tests/test_account_move.py [hunk 1]
```diff
@@ -298,3 +298,45 @@ def test_product_valuation_method_change_to_automated_negative_on_hand_qty(self)
...
+        at the creation of the account move with sub company
+        """
+        main_company = self.env.company
+        branch_a = self.env['res.company'].create({
+            'name': 'Branch A',
+            'country_id': main_company.country_id.id,
+            'parent_id': main_company.id,
+        })
+
+        self.env.user.write({'company_ids': (main_company + branch_a).ids, 'company_id': branch_a.id})
```
Comment (adwid) [inline addons/stock_account/tests/test_account_move.py:314]:
```suggestion
        self.env.user.company_id: branch_a
```

File: addons/stock_account/tests/test_account_move.py [hunk 2]
```diff
@@ -298,3 +298,45 @@ def test_product_valuation_method_change_to_automated_negative_on_hand_qty(self)
...
+
+        self.cr.precommit.run()
```
Comment (adwid) [inline addons/stock_account/tests/test_account_move.py:316]: ?

I guess it's for the CoA ? It's worth adding a comment with such mysterious lines
That being said, I would rather call `try_loading`, so we avoid the heavy mechanism of the `precommit`

You could also move that in, at least, `TestAccountMoveStockCommon` ? As I'm sure it won't be the last bug with branches... I let you decide for that point

File: addons/stock_account/tests/test_account_move.py [hunk 3]
```diff
@@ -298,3 +298,45 @@ def test_product_valuation_method_change_to_automated_negative_on_hand_qty(self)
...
+
+        categ = self.auto_categ.copy({'property_cost_method': 'average', 'property_valuation': 'real_time'})
```
Comment (adwid) [inline addons/stock_account/tests/test_account_move.py:318]: Let's avoid a record creation
```suggestion
        self.auto_categ.write({'property_cost_method': 'average', 'property_valuation': 'real_time'})
```

File: addons/stock_account/tests/test_account_move.py [hunk 4]
```diff
@@ -298,3 +298,45 @@ def test_product_valuation_method_change_to_automated_negative_on_hand_qty(self)
...
+        sm = self.env['stock.move'].create({
+            'name': product.name,
+            'product_id': product.id,
+            'product_uom_qty': 1,
+            'product_uom': product.uom_id.id,
+            'location_id': in_picking.location_id.id,
+            'location_dest_id': in_picking.location_dest_id.id,
+            'picking_id': in_picking.id,
+        })
+        self.assertTrue(in_picking.button_validate())
```
Comment (adwid) [inline addons/stock_account/tests/test_account_move.py:341]: This one is a bit useless
If you want to ensure that the picking is processed, you could do this
```suggestion
        in_picking.button_validate()
        self.assertEqual(sm.state, 'done')
```

File: addons/stock_account/tests/test_account_move.py [hunk 5]
```diff
@@ -298,3 +298,45 @@ def test_product_valuation_method_change_to_automated_negative_on_hand_qty(self)
...
+        self.assertEqual(self.env.company, sm.account_move_ids.company_id)
```
Comment (adwid) [inline addons/stock_account/tests/test_account_move.py:342]: The first value should always be the tested one, and the second value the expected one
```suggestion
        self.assertEqual(sm.account_move_ids.company_id, self.env.company)
```

**Review by adwid — COMMENTED** [review]
The ORM will detect a company inconsistency since the product belongs to the branch, while the AMLs are related to the main company

When computing the company of the AM (and therefore, the one of the AML), it seems we give the priority to the company of the journal before the one of the `env`
https://github.com/odoo/odoo/blob/90b1510d858fcdf9fb81572a81ab5ae3cff73d0c/addons/account/models/account_move.py#L694-L698

Hence your fix

But, with your diff, the AM will always be linked to the branch, while the bug here occurs because of the branch-specific product.
What if the product is now a generic one ? Do we still want the AM to be linked to the branch ?
@thomasdeleval Can we have your opinion here ?

File: addons/stock_account/tests/test_account_move.py (hunk 5 above)
Comment (adwid) [reply addons/stock_account/tests/test_account_move.py:342]: Or even
```suggestion
        self.assertEqual(sm.account_move_ids.company_id, branch_a)
```
To make it really clear

**Comment by guva-odoo** [comment]
> But, with your diff, the AM will always be linked to the branch, while the bug here occurs because of the branch-specific product. What if the product is now a generic one ? Do we still want the AM to be linked to the branch ?

If we have a generic product, we have the same error with svl here https://github.com/odoo/odoo/blob/d4ffcf8e8b260637c53d99e9b839132d91898f57/addons/stock_account/models/stock_move.py#L296

File: addons/stock_account/tests/test_account_move.py [hunk 6]
```diff
@@ -47,6 +47,24 @@ def setUpClass(cls, chart_template_ref=None):
...
+
+    @classmethod
+    def create_comp(cls):
+        comp = cls.env['res.company'].create({
+            'name': 'Company A',
+        })
+
+        cls.env['account.chart.template'].try_loading('generic_coa', comp, install_demo=False)
+        return comp
+
```
Comment (adwid) [inline addons/stock_account/tests/test_account_move.py:67]: Can you remove both ?
The branch method is not that useful, the company one is not even used and a better version already exists (`setup_company_data`)

File: addons/stock_account/tests/test_account_move.py [hunk 7]
```diff
@@ -298,3 +316,38 @@ def test_product_valuation_method_change_to_automated_negative_on_hand_qty(self)
             [stock_valuation_line, output_line],
             [expected_valuation_line, expected_output_line]
         )
+
+    def test_stock_account_move_automated_not_standard_with_branch_company(self):
+        """
+        Test that the validation of a stock picking does not fail `_check_company`
+        at the creation of the account move with sub company
+        """
+        branch_a = self.create_branch()
```
Comment (adwid) [inline addons/stock_account/tests/test_account_move.py:325]: Can you move the branch creation in the setup ?
So anyone can use it if needed

File: addons/stock_account/tests/test_account_move.py (hunk 6 above)
Comment (guva-odoo) [reply addons/stock_account/tests/test_account_move.py:67]: I don't even remember why I pushed `create_comp`

File: addons/stock_account/tests/test_account_move.py [hunk 8]
```diff
@@ -47,6 +47,17 @@ def setUpClass(cls, chart_template_ref=None):
             }
         )
 
+        cls.branch_a = cls.setup_company_data("Branch A", parent_id=cls.env.company.id)
+
+    @classmethod
+    def create_branch(cls, name=None, parent_company=None):
```
Comment (adwid) [inline addons/stock_account/tests/test_account_move.py:53]:
```suggestion
```

File: addons/stock_account/models/stock_move.py [hunk 9]
```diff
@@ -550,6 +550,7 @@ def _prepare_account_move_vals(self, credit_account_id, debit_account_id, journa
             'stock_valuation_layer_ids': [(6, None, [svl_id])],
             'move_type': 'entry',
             'is_storno': self.env.context.get('is_returned') and self.env.company.account_storno,
+            'company_id': self.env.company.id,
```
Comment (adwid) [inline addons/stock_account/models/stock_move.py:553]:
```suggestion
            'company_id': self.company_id.id,
```
This would be better actually

Comment (guva-odoo) [reply addons/stock_account/models/stock_move.py:553]: Are you sure? We do exactly `self.env.company` the. line above, and would it be the same in any case?

Comment (adwid) [reply addons/stock_account/models/stock_move.py:553]: Damn, good catch
Can you change that one too ?

**Comment by adwid** [comment]
robodoo r+

**Pull request closed without merging**